blog/content/post/2021-10-10-optiplex-dasharo.md: add new post (#438)
3mdeb/news-and-ideas: blog/content/post/2021-10-10-optiplex-dasharo.md

---
title: 'Dasharo for Dell OptiPlex 7010 / 9010'

> abstract: 'Open source firmware may be hundred times better than the

Reviewer: ```suggestion
abstract: 'Open source firmware may be hundreds of times better than the
```

           proprietary one. On the example of Dell OptiPlex 7010 / 9010 we will
           show you the advantages of Dasharo firmware on this machine.'
cover: /covers/dasharo-sygnet.svg
author: michal.zygowski
layout: post
published: true
date: 2021-10-10
archives: "2021"

tags:
  - firmware
categories:
  - Firmware

---

# Introduction

Dell OptiPlex 7010 / 9010 is the majority of machines used in 3mdeb office by the

> engineers and developers. It is only natural for a company propagating open

Reviewer 2: ```suggestion
engineers and developers. It is only natural for a company promoting open
```

source firmware (OSFW) to utilize their own product in daily work. The process
of porting coreboot on the machine took significant amount of free time (which

> you may read about in [other blog posts](https://blog.3mdeb.com/2020/2020-06-24-dell-optiplex-port/)

Reviewer 2: @miczyg1 maybe we can point to tag that mark both optiplex posts?

Author: Added new tags to the Optiplex posts and now the URL should link to them all

Reviewer 2: Ok, in long run maybe consistency between Dasharo, coreboot and our blog post, naming could be achieved.

by the way) but it was definitely worth it. It opened a path for developing a
firmware that would squeeze out the full potential of the platform. You may be
wondering what else a platform may do that OEM didn't implement. But I will
tell you this, with open source firmware almost anything is possible and with
[Dasharo firmware](https://dasharo.com/) we make it a reality.

## ME neutralization

Intel Management Engine is a very controversial component of modern Intel
chipsets and SoCs due to its capabilities and features. One of the most fearful
features is the remote management and connectivity. Intel ME is able to utilize
network via dedicated network controller on vPro machines without any awareness
of the operating system software. It is used for software like Intel Active
Management Technology ([AMT](https://www.intel.co.uk/content/www/uk/en/architecture-and-technology/intel-active-management-technology.html))

> or [Computrace](https://i.dell.com/sites/content/business/solutions/brochures/en/Documents/absolute-overview.pdf).

Author: i don't know how, but I found references that Computrace is using Intel Anti-Theft Technology (AT-p) and  AT-p is a service runner by ME. This [whitepaper](https://media9.connectedsocialmedia.com/intel/06/4470/Intel_Anti_Theft_Technology_Computrace_WhitePaper.pdf) says:
```
Computrace,* a leading IT asset-management and security solution from Absolute® Software,  now takes advantage of Intel® Anti-Theft Technology – PC Protection (Intel® AT-p).
```
And here is a photo of an app that checks ME Management and Security Status showing AT-p as a ME service:
![scr01](https://user-images.githubusercontent.com/30656491/140996658-38157c4d-f7d0-4d1f-8d37-4950c1eae4bc.png)
For me this is enough to prove the relation

> or [Computrace](https://i.dell.com/sites/content/business/solutions/brochures/en/Documents/absolute-overview.pdf).

Reviewer 2: @miczyg1 maybe providing link to that information, most probably using [wayback machine](https://web.archive.org/web/2020*/https://media9.connectedsocialmedia.com/intel/06/4470/Intel_Anti_Theft_Technology_Computrace_WhitePaper.pdf) in case of document disappearing in long run,  would be good addition to blog post.

> or [Computrace](https://i.dell.com/sites/content/business/solutions/brochures/en/Documents/absolute-overview.pdf).

Author: This page says it is discontinued for AT 1.x and 2.x, it doesn't say anything about newer versions thou @XenoKovah 

@pietrushnic again, that was not my point to elaborate on proofs here.

> or [Computrace](https://i.dell.com/sites/content/business/solutions/brochures/en/Documents/absolute-overview.pdf).

Author: Added the link to the picture and reference to the whitepaper.

Such features and capabilities often raises doubt and concerns about privacy.
Moreover it is a few megabytes of proprietary closed source firmware of unknown
quality with uncertain other impacts on the system. Many people often want to
disable ME or even neutralize it to reduce the attack surface. This is what we
offer with Dasharo firmware on Dell OptiPlex 7010 / 9010, [me_cleaner](https://github.com/corna/me_cleaner)
is being applied on the firmware image which can be flashed on the board. From

> a 6MB of unknown firmware only dozens of kilobytes are left.

Reviewer 2: I think referring this would be great: https://osfw.slack.com/archives/C9ZLS0U4F/p1633701873113300

Author: Added two sentences. @pietrushnic Please see if this is what you had in mind

Reviewer 2: @miczyg1 I think last sentence you added that blog posts are contradicting should be changed. In EU we have clear article talking about decompilation (article 6): https://eur-lex.europa.eu/legal-content/EN/TXT/?uri=CELEX%3A32009L0024

Finally you can say IANAL

Author: Done, last sentence removed and added reference to the decompilation article.

## RAM support

As engineers we often encounter situation that the same board comes in a few
SKUs (stock keeping units) which may be distinguished with by the used
processor for example. Such a procedure is often utilized to reduce the cost of
the product or rather offer a product less or more equipped to make it more
affordable on the market. Also as engineers doing open source firmware we want
to know how many differences they are, whether we may unify support of the
variants or even use the same binary to support multiple boards. But in the
cases where the mainboard is essentially identical and the OEM offers it with
different hardware capabilities, one question comes to the mind immediately:
what is going on? It happens for the 7010 and 9010 variants in the [specsheet](https://www.dell.com/support/manuals/en-us/optiplex-7010/opti7010_usff/specifications?guid=guid-157e8495-34d3-4efa-ab61-1d9efba4c90e).

```
Maximum memory:
  OptiPlex 7010	16 GB
  OptiPlex 9010	32 GB
```

Why would maximum RAM be limited by the variant? The number of memory slots and
the used processors is the same so there should not be any differences. It may

> be that the firmware limits the RAM capacity on the 7010 machine (very ugly)

Reviewer 2: @miczyg1 did we proved it?

Author: Hard to prove, but easy to check... However, we didn't do it yet.

Reviewer 2: Let's do it, maybe TAT can help.

Author: Still not checked, I may quickly find it out in the office tomorrow.

Author: Done the checks. 32GB RAM works on 7010 as well, no 16GB limitations or whatsoever.

one caveat. The Intel Q77 chipset used on the machine doesn't support CPUs with
ECC (currently only Xeon CPUs and some embedded ones support ECC). The CPU
compatibility of the chipset may be checked on [Intel ARK](https://ark.intel.com/content/www/us/en/ark/products/64027/intel-q77-express-chipset.html).
But no risk no fun. A decent Xeon CPU matching the board could be bought for
less than $100. So we did a quick test and... [it worked!](https://twitter.com/Dasharo_com/status/1435161914896748547?s=20).
We have a ECC capable CPU, now all we need is a ECC DIMM. Be aware though. A
typical server memory will not work here. We need an unbuffered ECC DIMM with
lower latencies (registered memory will only work on servers). However this
time it didn't work.

According to the investigations of the coreboot logs the CPU did not report the
ECC capability. The real question is why? There are a few pointers:

- [something else is controlling the capability](https://github.com/coreboot/coreboot/blob/master/src/northbridge/intel/sandybridge/raminit_common.c#L356)
- ECC support does not depend solely on CPU (the memory controller is a part of

>   the CPU)

Reviewer 2: @miczyg1 Very well played!

> In some depths of a sensitive black hole I have found an information hat the

Reviewer 2: ```suggestion
In some depths of a sensitive black hole I have found an information that the
```

perform trusted measurement of the operation system software defined in Trusted
Computing Group [D-RTM architecture specification](https://trustedcomputinggroup.org/wp-content/uploads/TCG_D-RTM_Architecture_v1-0_Published_06172013.pdf).
Although Dell firmware supports TXT and it is nothing new for Dasharo to
support it as well, we haven't yet been able to successfully launch TXT with
open source firmware on the Dell OptiPlex 7010 / 9010. But we intend to resolve
this issue. Moreover we aim to support [TrenchBoot](https://trenchboot.org/)
on this machine. A very good use case of Intel TXT was presented by Qubes OS
[Anti Evil Maid (AEM)](https://github.com/QubesOS/qubes-antievilmaid/). By the
way we have been working on enabling AMD CPUs and TPM 2.0 as well in AEM.
In an ideal case Dell OptiPlex will come with TXT and TrenchBoot support and
use AEM to securely measure Qubes OS and perform a remote attestation using
[Fobnail](https://fobnail.3mdeb.com/). To protect the user even further we also
plan to integrate [Intel STM](https://software.intel.com/content/www/us/en/develop/articles/smi-transfer-monitor-stm.html)

> to avoid System Management Mode attacks to which Intel TXT is vulnerable.

Reviewer 2: @miczyg1 there is a lot of very good meat here, I like it.

## Open Security Training 2

All these possibilities, open source firmware and security features makes the
Dell OptiPlex 7010 / 9010 a first class citizen secure workstation that has been
recognized by [Xeno Kovah](https://twitter.com/XenoKovah), a former employee of
Apple who worked in Apple's security department. Dell OptiPlex 7010 / 9010 is the
reference machine used in the [Open Security Training 2](https://ost2.fyi/) - a

> freely available firmware and security training, second edition.

Reviewer: Well, to say it's "a first class citizen secure workstation that has been recognized by" me, would imply that I had done an audit of the security and found no problems. Since I haven't, perhaps a better phrasing would be:

"
All these possibilities, open source firmware and security features makes the
Dell OptiPlex 7010 / 9010 a first class citizen secure workstation."

That then sounds more like your assertion, rather than mine. Followed by
"
It has been recognized by [Xeno Kovah](https://twitter.com/XenoKovah), a former Apple firmware security architect, as an excellent training platform for teaching Intel firmware security and coreboot courses. The Dell OptiPlex 7010 / 9010 is used as a reference machine in the [OpenSecurityTraining2](https://ost2.fyi/) "Architecture 4001: x86-64 Reset Vector Firmware" class (which is currently in private beta testing).
"

(We can update this with a link to the class later when it's out of beta if so desired.)

"
And that class then leads into the ["Architecture 4031: x86-64 Reset Vector: coreboot"](https://p.ost2.fyi/courses/course-v1:OpenSecurityTraining2+Arch4031_x86-64_RV_coreboot+2021_v1/about) class, and future coreboot classes which are under development, such as "Architecture 4032: coreboot Hardware Hands-On".
"

Author: Done. @XenoKovah please check if the form suits you now.

Computing Group [D-RTM architecture specification](https://trustedcomputinggroup.org/wp-content/uploads/TCG_D-RTM_Architecture_v1-0_Published_06172013.pdf).
Although Dell firmware supports TXT and it is nothing new for Dasharo to
support it as well, we haven't yet been able to successfully launch TXT with
open source firmware on the Dell OptiPlex 7010 / 9010. But we intend to resolve
this issue. Moreover we aim to support [TrenchBoot](https://trenchboot.org/)
on this machine. A very good use case of Intel TXT was presented by Qubes OS
[Anti Evil Maid (AEM)](https://github.com/QubesOS/qubes-antievilmaid/). By the
way we have been working on enabling AMD CPUs and TPM 2.0 as well in AEM.
In an ideal case Dell OptiPlex will come with TXT and TrenchBoot support and
use AEM to securely measure Qubes OS and perform a remote attestation using
[Fobnail](https://fobnail.3mdeb.com/). To protect the user even further we also
plan to integrate [Intel STM](https://software.intel.com/content/www/us/en/develop/articles/smi-transfer-monitor-stm.html)
to avoid System Management Mode attacks to which Intel TXT is vulnerable.

## Open Security Training 2

> All these possibilities, open source firmware and security features makes the

Author: Done
